external/llvm-project/mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp

```diff
@@ -148,6 +164,30 @@ void SerializeToHsacoPass::getDependentDialects(
   gpu::SerializeToBlobPass::getDependentDialects(registry);
 }
 
+LogicalResult
+SerializeToHsacoPass::optimizeLlvm(llvm::Module &llvmModule,
+                                   llvm::TargetMachine &targetMachine) {
```

Comment:
Should you be calling `gpu::SerializeToBlobPass::optimizeLlvm()`? It is a bit unclear to me why SearializeToBlobPass needs to be Updated?

Reply:
I guess I'm just confused that on the scope of this PR I don't see `SerializeToBlobPass` get pulled in any of our stuff, we still use `SearilizeToHsaco` only. 

> so that we didn't need to recreate the rest of MLIR -> ISA pipeline in our subclass

Sorry, could you elaborate on this a bit more. I didn't quite catch what the problem would be if you don't modify SerializeToBlob pass.

Reply:
`SerializeToBlob` is a superclass of `SerializeToHsaco`.

The function `translateToISA` I modified is a member of SerializeToBlob, not SerializeToHsaco, and SerializeToHsaco doesn't (and, I argue, shouldn't) override it in its entirety. The current hotfix actually patches `translateToISA` inline to add the optimizations.

In order to not have a hacky patch to a generic class, I've added an `optimizeLlvm()` method that subclasses of `SerializeToBlob` can overload.

Reply:
> SerializeToBlob is a superclass of SerializeToHsaco.

Didn't catch this, but this make sense for now. And the way you do the patch through `optimizeLlvm()` looks like a reasonable way to do it: So that whatever they choose to do with this function will have minimal impact on us in the future.